### PR TITLE
[model] throw exception when airqualitystate gets values smaller than 0

### PR DESCRIPTION
**Pull request by michelececcacci**

### CODE REVIEW

**Comment by aleemont1** [comment]
Don't we have the Limit class for this purpose? I didn't quite looked at that but I thiught it was meant to be used for this.

**Comment by michelececcacci** [comment]
well, `limit` currently needs both an upper and lower bound. we can merge this as is, or use `optional` in the limit class.

**Comment by aleemont1** [comment]
> well, `limit` currently needs both an upper and lower bound. we can merge this as is, or use `optional` in the limit class.

Yup you're right, `Optional` might be a good idea @smagnaniuni

**Review by aleemont1 — COMMENTED** [review]
Just the Limit thing.

**Comment by smagnaniuni** [comment]
If you mean using optional in parameters that is a code smell, it's much better to just create two other methods: lowerLimit/lowerClamp, upperLimit/upperClamp for the purpose

**Comment by michelececcacci** [comment]
>If you mean using optional in parameters that is a code smell, it's much better to just create two other methods: lowerLimit/lowerClamp, upperLimit/upperClamp for the purpose

well, at this point it would be the `max`/`min` situation all over again. How about using a builder pattern in limit's constructor?

**Comment by smagnaniuni** [comment]
Explain what you mean, I am not sure what you mean with the max/min.

**Comment by michelececcacci** [comment]
I don't see the difference between `LowerClamp` and `math.min `. What i was proposing was adding an `AddLowerLimit` and `AddUpperLimit` to `Limit`

**Comment by smagnaniuni** [comment]
At the moment there is no need to create an instance and I don't see the need to, the builder pattern is overkill.
The difference is that min, max only support specific primitive types.

**Comment by michelececcacci** [comment]
> At the moment there is no need to create an instance and I don't see the need to, the builder pattern is overkill. The difference is that min, max only support specific numeric types.

ok, this is fine. but i would still prefer to merge this, since limit doesn't throw exceptions, and in this specific case it's useful.

**Comment by smagnaniuni** [comment]
If you want to build upon the limit class, there is also the `Bounds` class which could serve the purpose of the `Limit::clamp` method, and then you could create two other classes `LowerBound` `UpperBound` which have their own limit/clamp method, and then build the `Bounds` object with those two. But this also sounds like an over complication for something that only serves 1 purpose and is not complex.

**Comment by michelececcacci** [comment]
seems overkill to me

File: src/main/java/homer/model/airquality/AirQualityStateImpl.java [hunk 1]
```diff
@@ -18,7 +18,11 @@ public final class AirQualityStateImpl implements AirQualityState {
      * @param pm25
      */
     public AirQualityStateImpl(final double co2, final double pm10, final double toxicGasPercentage, final double pm25) {
-        this.co2 = co2;
+        requireGreaterOrEqualThanZero(co2); 
+        requireGreaterOrEqualThanZero(pm10); 
+        requireGreaterOrEqualThanZero(toxicGasPercentage);
+        requireGreaterOrEqualThanZero(pm25); 
+        this.co2 =  co2;
```
Comment (smagnaniuni) [inline src/main/java/homer/model/airquality/AirQualityStateImpl.java:25]: lgtm, if you want to save some lines you could use the setters in the constructor

**Review by smagnaniuni — APPROVED** [review]
lgtm

File: src/main/java/homer/model/airquality/AirQualityStateImpl.java (hunk 1 above)
Comment (michelececcacci) [reply src/main/java/homer/model/airquality/AirQualityStateImpl.java:25]: that's right

**Pull request merged**